File: TypeScript/TP3_MET_TS/src/Inductif/MotMotInductif.ts
import {IterateurMot} from "../Interface/IterateurMot";
import {MotVideInductif} from "./MotVideInductif";
import {CaractereMotInductif} from "./CaractereMotInductif";
import {Mot} from "../Interface/Mot";

export class MotMotInductif extends Mot {

    public FAB: Mot = MotVideInductif.SINGLETON;

    private gauche: Mot;
    private droite: Mot;

    public constructor(gauche: Mot, droite: Mot) {
        super();
        this.gauche = gauche;
        this.droite = droite;
    }

    public creerMotVide(): Mot {
        return MotVideInductif.SINGLETON;
    }

    public casMotMot(): boolean {
        return true;
    }

    public caractere(): string {
        let i: IterateurMot = this.iterateurInductif();
        return i.next();
    }

    public reste(): Mot {
        let i: IterateurMot = this.iterateurInductif();
        return i.reste();
    }

    public getMotGauche(): Mot {
        return this.gauche;
    }

    public getMotDroite(): Mot {
        return this.droite;
    }

    public creerCaractereMot(c: string): Mot {
        return new CaractereMotInductif(c, this);
    }

    public creerMotMot(droite: Mot): Mot {
        return new MotMotInductif(this, droite);
    }
}

File: Java/TP3_MET_JAVA/src/Inductif/MotMotInductif.java
package Inductif;

import Interface.IterateurMot;
import Interface.Mot;

public class MotMotInductif implements Mot {

    public static final Mot FAB = MotVideInductif.SINGLETON;

    private Mot gauche;
    private Mot droite;

    public MotMotInductif(Mot gauche, Mot droite) {
        this.gauche = gauche;
        this.droite = droite;
    }

    @Override
    public Mot creerMotVide() {
        return MotVideInductif.SINGLETON;
    }

    @Override
    public boolean casMotMot() {
        return true;
    }

    @Override
    public Character caractere() {
        IterateurMot i = this.iterateurInductif();
        return i.next();
    }

    @Override
    public Mot reste() {
        IterateurMot i = this.iterateurInductif();
        return i.reste();
    }

    @Override
    public Mot getMotGauche() {
        return this.gauche;
    }

    @Override
    public Mot getMotDroite() {
        return this.droite;
    }

    @Override
    public Mot creerCaractereMot(char c) {
        return new CaractereMotInductif(c, this);
    }

    @Override
    public Mot creerMotMot(Mot droite) {
        return new MotMotInductif(this, droite);
    }

}
